controlplane/kubeadm/internal/etcd/etcd.go

```diff
@@ -0,0 +1,233 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd
+
+import (
+	"context"
+	"crypto/tls"
+	"net"
+
+	"github.com/pkg/errors"
+	"go.etcd.io/etcd/clientv3"
+	"go.etcd.io/etcd/etcdserver/etcdserverpb"
+	"google.golang.org/grpc"
+)
+
+// GRPCDial is a function that creates a connection to a given endpoint.
+type GRPCDial func(ctx context.Context, addr string) (net.Conn, error)
+
+// etcd wraps the etcd client from etcd's clientv3 package.
+// This interface is implemented by both the clientv3 package and the backoff adapter that adds retries to the client.
+type etcd interface {
+	AlarmList(ctx context.Context) (*clientv3.AlarmResponse, error)
+	Close() error
+	Endpoints() []string
+	MemberList(ctx context.Context) (*clientv3.MemberListResponse, error)
+	MemberRemove(ctx context.Context, id uint64) (*clientv3.MemberRemoveResponse, error)
+	MemberUpdate(ctx context.Context, id uint64, peerURLs []string) (*clientv3.MemberUpdateResponse, error)
+	MoveLeader(ctx context.Context, id uint64) (*clientv3.MoveLeaderResponse, error)
+	Status(ctx context.Context, endpoint string) (*clientv3.StatusResponse, error)
+}
+
+// Client wraps an etcd client formatting its output to something more consumable.
+type Client struct {
+	EtcdClient etcd
+	Endpoint   string
+}
+
+// MemberAlarm represents an alarm type association with a cluster member.
+type MemberAlarm struct {
+	// MemberID is the ID of the member associated with the raised alarm.
+	MemberID uint64
+
+	// Type is the type of alarm which has been raised.
+	Type AlarmType
+}
+
+type AlarmType int32
+
+const (
+	// AlarmOK denotes that the cluster member is OK.
+	AlarmOk AlarmType = iota
+
+	// AlarmNoSpace denotes that the cluster member has run out of disk space.
+	AlarmNoSpace
+
+	// AlarmCorrupt denotes that the cluster member has corrupted data.
+	AlarmCorrupt
+)
+
+// Adapted from kubeadm
+
+// Member struct defines an etcd member; it is used to avoid spreading
+// github.com/coreos/etcd dependencies.
+type Member struct {
+	// ID is the ID of this cluster member
```

Comment:
Every member of the same cluster has the same `ClusterID`, which is included in the standard response headers of an etcd API request. It would help to record this value here, for example for any consumer that wants to easily verify that multiple members belong to the same cluster.

The value is defined by the first etcd member, and never changes.

Reply:
ooh, nice. That could be much simpler than checking the peer list.

Reply:
Are you suggesting we make this a stateful object/cache api responses?